### PR TITLE
Introduce conditional_requirements field in commands.json

### PR DESCRIPTION
**Pull request by bashi**
Context: https://github.com/web-platform-tests/wpt/pull/29727#discussion_r695287300

I'd like to introduce aioquic dependency to `wpt run` (and `wpt serve` later) using virtualenv for WebTransport tests. This PR introduces a new field `optional_requirements` in commands.json so that it can specify additional requirements when a command line flag is provided.

Alternatively, we may just add "../webtransport/requirements.txt" into `requirements` field of tools/wpt/commands.json. This unconditionally installs aioquic when we run `wpt run`, but per https://github.com/web-platform-tests/rfcs/pull/85 it may be acceptable.

### CODE REVIEW

**Comment by bashi** [comment]
@jgraham @foolip What do you think about this? I'm happy to create a RFC if needed.

@yutakahirano @nidhijaju FYI

**Review by jgraham — COMMENTED** [review]
I like this! We could in theory generalise it to allow supporting the wptrunner use case for installing per-product requirements like `{"optional_requirements": {"product:chrome": [requirements_chrome.txt]}}`.

Also, thanks for taking the time to add documentation, it's much appreciated.

I do think a quick RFC would be appropriate for this kind of change.

**Comment by bashi** [comment]
I'm glad to hear you like the idea :) I'll create a RFC, maybe tomorrow JST.

**Comment by bashi** [comment]
Updated based on https://github.com/web-platform-tests/rfcs/pull/96. @jgraham @foolip  Could you take another look?

File: tools/wptrunner/docs/commands.rst [hunk 1]
```diff
@@ -61,21 +61,26 @@ additional fields. All paths are relative to the commands.json.
   running the subcommand. :code:`virtualenv` must be true when this field is
   set.
 
-:code:`optional_requirements`
-  Similar to :code:`requirements`, but requirements specified in this list are
-  installed only when the corresponding command line flag is specified for the
-  subcommand. For example, given the following commands.json::
+:code:`conditional_requirements`
+  A key-value object. Values in this object are lists of paths similar to
```
Comment (yutakahirano) [inline tools/wptrunner/docs/commands.rst:65]: [optional] How about the following?

A key-value object. Each key represents a condition, and value represents additional requirements when the condition is met. The requirements have the same format as :code:`requirements`. Currently "commandline_flag" is...

**Review by yutakahirano — APPROVED** [review]
LGTM

File: tools/wptrunner/docs/commands.rst (hunk 1 above)
Comment (bashi) [reply tools/wptrunner/docs/commands.rst]: That's much nicer! Done.

**Comment by bashi** [comment]
Friendly ping.

**Comment by foolip** [comment]
From the logs the Firefox failure reminds me a lot of https://github.com/web-platform-tests/wpt/issues/30796. I'm going to close and reopen the PR and see if it passes, in which case it's probably flaky with the same root cause.

**Comment by foolip** [comment]
Yep, it passed now, so looks like flakiness. Please follow https://github.com/web-platform-tests/wpt/issues/30796 for investigation.

File: tools/wptrunner/docs/commands.rst [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    "foo": {
+      "path": "foo.py",
+      "script": "run",
+      "parser": "get_parser",
+      "help": "Run foo"
+    },
+    "bar": {
+      "path": "bar.py",
+      "script": "run",
+      "vritualenv": true,
```
Comment (foolip) [inline tools/wptrunner/docs/commands.rst:19]:
```suggestion
      "virtualenv": true,
```

File: tools/wptrunner/docs/commands.rst [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+:code:`help`
+  Brief description of the subcommand.
+
+:code:`virtualenv`
+  When True, the subcommand is executed with a virtualenv environment. Default
+  to True.
+
+:code:`install`
+  A list of strings of each string represents a
```
Comment (foolip) [inline tools/wptrunner/docs/commands.rst:52]:
```suggestion
  A list of strings where each string represents a
```

File: tools/wptrunner/docs/commands.rst [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  `requirement object <https://setuptools.readthedocs.io/en/latest/pkg_resources.html#requirement-objects>`_.
+  These requirements are installed into the virtualenv environment before
+  running the subcommand. :code:`virtualenv` must be true when this field is
+  set.
+
+:code:`requirements`
+  A list of paths of each path specifies a requirements.txt. All requirements
```
Comment (foolip) [inline tools/wptrunner/docs/commands.rst:59]:
```suggestion
  A list of paths where each path specifies a requirements.txt. All requirements
```

Comment (bashi) [reply tools/wptrunner/docs/commands.rst]: Done

File: tools/wptrunner/docs/commands.rst (hunk 3 above)
Comment (bashi) [reply tools/wptrunner/docs/commands.rst]: Done

File: tools/wptrunner/docs/commands.rst (hunk 2 above)
Comment (bashi) [reply tools/wptrunner/docs/commands.rst:19]: Done

File: tools/wpt/wpt.py [hunk 5]
```diff
@@ -14,6 +14,28 @@
...
 
+def load_conditional_requirements(props, base_dir):
+    """Load conditional requirements from commands.json."""
+
+    conditional_requirements = props.get("conditional_requirements", {})
+    if not conditional_requirements:
+        return {}
+
+    commandline_flag_requirements = {}
+    for key, value in props.get("conditional_requirements", {}).items():
```
[the remark this answers is not in the document]
Comment (bashi) [reply tools/wpt/wpt.py]: Good catch! Done.

File: tools/wpt/wpt.py [hunk 6]
```diff
@@ -14,6 +14,28 @@
 wpt_root = os.path.abspath(os.path.join(here, os.pardir, os.pardir))
 
 
+def load_conditional_requirements(props, base_dir):
+    """Load conditional requirements from commands.json."""
+
+    conditional_requirements = props.get("conditional_requirements", {})
+    if not conditional_requirements:
```
[the remark this answers is not in the document]
Comment (bashi) [reply tools/wpt/wpt.py:21]: Yes, as https://docs.python.org/3/library/stdtypes.html#truth-value-testing says, an empty dict is falsy in Python. I agree that it's not obvious. Changed to use default get() which returns None when the key doesn't exist IIUC.

**Comment by bashi** [comment]
Thank you @foolip for review and the link to flakiness!

I'm going to merge this PR soon-ish.

**Pull request merged**